numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -423,150 +424,295 @@ gentype_format(PyObject *self, PyObject *args)
+ *       'k': keep trailing zeros up to prec, and keep '.'
+ *      '\0': trim trailing zeros and '.'
+ *       '.': trim zeros but leave "."
+ *       '0': trim zeros but leave ".0".
+ *            Important: input string must already contain the .0 to use this option
+ *       'r': like '\0' for scientific notation, like '0' otherwise
+ *            (this emulates python's repr(float) behavior)
+ */
+NPY_NO_EXPORT void
+trim_float_gformat_zeros(char *repr, char tail)
+{
+    int epos=0, nzpos;
+    while (repr[epos] != 'e' && repr[epos] != '\0') {
+        epos++;
+    }
+    nzpos = epos - 1;
+    while (nzpos > 0 && repr[nzpos] == '0') {
+        nzpos--;
+    }
+
+    if (repr[nzpos] == '.') {
+        /* in "r" mode, print ".0" for non-exponential ouput only */
+        if (tail == 'r') {
+            if (repr[epos] != 'e') {
+                nzpos += 2;
+            }
+        }
+        else if (tail == '.'){
+            nzpos += 1;
+        }
+        else if (tail == '0') {
+            nzpos += 2;
+        }
+    }
+    else {
+        /* we found a trailing nonzero-digit instead of '.' */
+        nzpos += 1;
+    }
+
+    if (nzpos == epos) {
+        return;
+    }
+
+    // copy exponent back to offset position
+    while (repr[epos] != '\0') {
+        repr[nzpos++] = repr[epos++];
+    }
+    repr[nzpos] = repr[epos];
+}
+
 #ifdef FORCE_NO_LONG_DOUBLE_FORMATTING
 #undef NPY_LONGDOUBLE_FMT
 #define NPY_LONGDOUBLE_FMT NPY_DOUBLE_FMT
 #endif
 
 /**begin repeat
- * #name = float, double, longdouble#
- * #NAME = FLOAT, DOUBLE, LONGDOUBLE#
- * #type = npy_float, npy_double, npy_longdouble#
- * #suff = f, d, l#
+ * #name = float, double#
+ * #NAME = FLOAT, DOUBLE#
+ * #type = npy_float, npy_double#
+ * #suff = f, d#
  */
 
-#define _FMT1 "%%.%i" NPY_@NAME@_FMT
-#define _FMT2 "%%+.%i" NPY_@NAME@_FMT
+/*
+ * Converts a floating point value to a string.
+ * Arguments are similar to those of PyOS_double_to_string:
+ *   * format_code: similar to the argument to PyOS_double_to_string,
+ *   * prec: same as the argument to PyOS_double_to_string
+ *   * sign: boolean value, controls whether sign is always printed
+ *   * tail: one of '\0', '.' or '0', to control what happens for
+ *           floats that look like integers. For '.', add a decimal
+ *           point after the number. For '0', add ".0" after the number.
+ *           For '\0', don't add anything. Has no effect for 'r' format.
+ */
+NPY_NO_EXPORT PyObject *
+format_@name@(@type@ val, char format_code, unsigned int prec, int sign,
+              char tail)
+{
+    char *repr;
+    PyObject *ret;
+    int flags = 0;
+
+    if (sign) {
+        flags |= Py_DTSF_SIGN;
+    }
 
-NPY_NO_EXPORT void
-format_@name@(char *buf, size_t buflen, @type@ val, unsigned int prec)
+    if (tail == 'r' || tail == '0') {
+        flags |= Py_DTSF_ADD_DOT_0;
+    }
+    else {
+        flags |= Py_DTSF_ALT;
+    }
+
+    /* 'g' format precision is 1 greater, so decr for consistency with f,e */
```

Comment:
I'm a bit confused why we do this if the goal is to be more similar to `PyOS_double_to_string`. Is it just that we do it elsewhere too?